### PR TITLE
Fixed #17454 and cleanup

### LINKED ISSUE #17454: ValueError: solve((1 - x - I)**4, x)
```
>>> solve((1 - x - I)**4, x)
Traceback (most recent call last):
  File "test.py", line 25, in <module>
    print(solve(e, x))
  File "/home/eward/se/sympy/solvers/solvers.py", line 1178, in solve
    solution = _solve(f[0], *symbols, **flags)
  File "/home/eward/se/sympy/solvers/solvers.py", line 1661, in _solve
    soln = roots(poly, **solvers)
  File "/home/eward/se/sympy/polys/polyroots.py", line 1017, in roots
    for currentroot in _try_decompose(f):
  File "/home/eward/se/sympy/polys/polyroots.py", line 915, in _try_decompose
    for currentroot in _try_heuristics(factors[0]):
  File "/home/eward/se/sympy/polys/polyroots.py", line 961, in _try_heuristics
    result += roots_quartic(f)
  File "/home/eward/se/sympy/polys/polyroots.py", line 315, in roots_quartic
    roots([1, e, g], multiple=True)]
ValueError: not enough values to unpack (expected 2, got 1)
```

Since [1, e, g] is a quadratic, `roots_quartic` is expecting two results on line 315. `roots` does not behave as expected if `e` is not expanded first:

```
>>> e = -12*I - 3*(-4 + 4*I)**2/8
>>> expand(e)
0
>>> roots([1, e, 0], multiple=True)
[0]
>>> roots([1, expand(e), 0], multiple=True)
[0, 0]
```

### PR DESCRIPTION
**Pull request by oscargus**
#### References to other Issues or PRs

Fixes #17454 

#### Brief description of what is fixed or changed
Adding multiple zero roots now works better. The main problem is that some expressions that evaluates to zero are not being detected to be zero at the start. Hence, one must add them as one goes along.

#### Other comments
Consider the value `a=-3*(-4 - 4*I)**2/8 + 12*I`

In the original code the test for zero was `a == S.Zero` which returns `False`.

Now, the test is `a.is_zero` which for some expressions evaluates to `True` where the above don't.

I addition, since `a.simplify()` gives `0`, there is a need to later on in the iteration be able to add these "late found" zeros.

#### Release Notes

* polys
    * Fixed issue with solving quartic polynomial with complex coefficients.
    * Improved multiple root finding when roots are complicated expressions evaluating to zero.

### CODE REVIEW

**Comment by ethankward** [comment]
What about the underlying issue where `roots` only returns 1 value?

**Comment by oscargus** [comment]
I've fixed that as well now. Turned out that the polynomial should have any zero roots after a while, but with "complicated" coefficients eventually simplifying to zero those were not properly stored.

**Comment by oscargus** [comment]
I removed the `_mexpand`s since it is the general problem with this. For some expressions it gets better (as the one above), but for some, originally simple expressions, it gets worse.

Still, a  bit of cleanup of redundant operations etc and the bug is fixed without the need to expand.

File: sympy/polys/polyroots.py [hunk 1]
```diff
@@ -902,7 +898,12 @@ def roots(f, *gens, **flags):
         if f.is_multivariate:
             raise PolynomialError('multivariate polynomials are not supported')
 
-    def _update_dict(result, currentroot, k):
+    def _update_dict(result, zeros, currentroot, k):
+        if currentroot == S.Zero:
+            if S.Zero in zeros:
+                zeros[S.Zero] += k
+            else:
+                zeros[S.Zero] = k
```
Comment (oscarbenjamin) [inline sympy/polys/polyroots.py]: Apparently this line isn't tested. It's not obvious to me how to hit it though...

Comment (jksuom) [reply sympy/polys/polyroots.py]: I don't think `zeros` should be modified here. All zero roots are extracted before anything else is done. New zero roots of modified polynomials may appear after translation/rescaling but those should not be added here as they will be non-zero after the transformations have been reverted.

**Comment by oscarbenjamin** [comment]
Like the associated issue #17454 this doesn't need to hold up the 1.5 release.

**Comment by czgdp1807** [comment]
It looks like the PR successfully fixes the associated issue. @oscargus If you don't mind, can I continue this in a new PR?

**Comment by czgdp1807** [comment]
ping @oscargus

**Comment by oscarbenjamin** [comment]
Note the comment here: https://github.com/sympy/sympy/pull/17482#discussion_r319374900

File: sympy/polys/polyroots.py (hunk 1 above)
Comment (oscargus) [reply sympy/polys/polyroots.py:906]: I do not recall exactly, but based on the comment https://github.com/sympy/sympy/pull/17482#issuecomment-523921526 I believe what happens is that one get pretty large expressions that SymPy cannot simplify to zero at the start. However, after some iterations it is realized that they actually are zero. So "All zero roots are extracted before anything else is done." is not how it works in practice as SymPy doesn't realize that they are zeros. Which is the root(!) cause of problem.

For the testing of that line, no idea. I either thought that there may be cases where `zeros[S.Zero]` is not initialized or found a case where it isn't while playing around.

File: sympy/polys/tests/test_polyroots.py [hunk 2]
```diff
@@ -710,3 +710,7 @@ def test_roots_composite():
 def test_issue_19113():
     eq = cos(x)**3 - cos(x) + 1
     raises(PolynomialError, lambda: roots(eq))
+
+
+def test_issue_17454():
+    assert roots([1, -3*(-4 - 4*I)**2/8 + 12*I, 0], multiple=True) == [0, 0]
```
Comment (oscargus) [inline sympy/polys/tests/test_polyroots.py:716]: It seems like the original issue was sorted out somehow, probably through better zero checking in e.g. https://github.com/sympy/sympy/pull/17510 but this still gives the incorrect result in the current master.

**Comment by oscargus** [comment]
I've updated the description and hopefully that clears your doubts @jksoum to why we need to add zeros as we go along.

**Comment by oscargus** [comment]
If nothing else, that allows 

`assert roots([1, -3*(-4 - 4*I)**2/8 + 12*I, 0], multiple=True) == [0, 0]` 

to pass.

**Pull request merged**